src/memray/_memray/compat.cpp

```diff
@@ -0,0 +1,40 @@
+#include "compat.h"
+
+namespace memray::compat {
+
+void
+setprofileAllThreads(Py_tracefunc func, PyObject* arg)
+{
+    assert(PyGILState_Check());
+
+    PyThreadState* this_tstate = PyThreadState_Get();
+    PyInterpreterState* interp = threadStateGetInterpreter(this_tstate);
+    for (PyThreadState* tstate = PyInterpreterState_ThreadHead(interp); tstate != nullptr;
```

Comment:
I think we may have a problem here. Is possible that calling `_PyEval_SetProfile` drops the GIL and that can do random stuff with threads? What's worse, `PyThreadState_Next` can be called at the same time a native extension destroys a thread state (this is a known CPython problem). 

I don't think we can do anything here until we have the new API I am working on

Reply:
> It's possible that calling `_PyEval_SetProfile` drops the GIL and that can do random stuff with threads, like creating new ones or destroying them (this may not be a problem, though).

`_PyEval_SetProfile` can only drop the GIL if either a) the `__del__` of the previous profile arg fires, or b) an audit hook is installed for `sys.setprofile`, right? We work around the first one with a call to `setprofileAllThreads(nullptr, nullptr)` before we try to do anything interesting (albeit it's possible that causes a user supplied `__del__` to run and call `sys.setprofile` itself, at least before 3.11, in which case we're screwed).

There's not much that we can do about the audit function running and dropping the GIL, though. Interestingly, that can't happen in 3.7 or 3.8 thanks to my backport of `_PyEval_SetProfile`...

Hm.

> What's worse, `PyThreadState_Next` can be called at the same time a native extension destroys a thread state

Ah. Yep. That's bad.

Reply:
But there's nothing we can do about it. Sounds like this will eventually be improved in CPython, and in the meantime we believe that this is unlikely to cause problems.